examples/threading/threading.c

```diff
@@ -0,0 +1,107 @@
+#include "threading.h"
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+
+// Optional: use these functions to add debug or error prints to your application
+#define DEBUG_LOG(msg,...) printf("DEBUG MSG: " msg "\n" , ##__VA_ARGS__)
+//#define DEBUG_LOG(msg,...) printf("threading: " msg "\n" , ##__VA_ARGS__)
+#define ERROR_LOG(msg,...) printf("threading ERROR: " msg "\n" , ##__VA_ARGS__)
+
+#define MILLI_SLEEP_TO_uSLEEP(x) (x * 1000)
+
+void* threadfunc(void* thread_param)
+{
+
+    // TODO: wait, obtain mutex, wait, release mutex as described by thread_data structure
+    // hint: use a cast like the one below to obtain thread arguments from your parameter
+    //struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+
+    struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+
+    bool ret_status = true;
+
+    // first sleep 
+    if(usleep(MILLI_SLEEP_TO_uSLEEP(thread_func_args->wait_to_obtain_ms))) {
+        // non zero value is failure
+        ret_status = false;
+
+        ERROR_LOG(" Wait to obtain sleep fail %d", errno);
+    
+    } else { // SUCCESS! now access mutex 
+        
+        if(pthread_mutex_lock(thread_func_args->mutex)) {
+            // non zero value is fail
+            ret_status = false;
+            ERROR_LOG(" Mutex acquire fail %d", errno);
+        
+        } else { // SUCCESS! now sleep... 
+
+            if(usleep(MILLI_SLEEP_TO_uSLEEP(thread_func_args->wait_to_release_ms))) {
+                
+                // non zero value is failure
+                ret_status = false;
+
+                ERROR_LOG(" Wait to release sleep fail %d", errno);
+                // unlock mutex, what if it fails? log error?
+                
+                if(pthread_mutex_unlock(thread_func_args->mutex)) {
+                    ERROR_LOG(" Mutex unlock fail %d", errno);
+                }
```

Comment:
This works but I don't think you need the code duplication with below, you could just always `pthread_mutex_unlock`